chunky/src/java/se/llbit/chunky/world/region/MCRegion.java

```diff
@@ -215,76 +223,102 @@ private ChunkDataSource getChunkData(ChunkPosition chunkPos) {
     File regionFile = new File(regionDirectory, fileName);
     ChunkDataSource data = null;
     if (regionFile.exists()) {
-      data = getChunkData(regionFile, chunkPos);
+      // TODO: reuse RandomAccessFile instances when loading world map
+      try (RandomAccessFile raf = new RandomAccessFile(regionFile, "r")) {
+        data = readChunkData(raf, chunkPos);
+      } catch (IOException ex) {
+        Log.warn(
+          String.format(
+            "Failed to read chunk %s in region %s",
+            chunkPos,
+            regionFile.getName()
+          ), ex
+        );
+      }
     }
     if (data == null) {
-      data = new ChunkDataSource((int) System.currentTimeMillis(), null);
+      data = new ChunkDataSource((int) System.currentTimeMillis());
     }
-    chunkTimestamps[(chunkPos.x & 31) + (chunkPos.z & 31) * 32] = data.timestamp;
+    chunkTimestamps[getMCAChunkIndex(chunkPos)] = data.timestamp;
     return data;
   }
 
   /**
    * Read chunk data from region file.
+   * <a href="https://wiki.vg/Region_Files#Structure">Format documentation</a>
    *
    * @return {@code null} if the chunk could not be loaded
    */
-  private static ChunkDataSource getChunkData(File regionFile, ChunkPosition chunkPos) {
-    int x = chunkPos.x & 31;
-    int z = chunkPos.z & 31;
-    int index = x + z * 32;
-    try (RandomAccessFile file = new RandomAccessFile(regionFile, "r")) {
-      long length = file.length();
-      if (length < 2 * SECTOR_SIZE) {
-        Log.warn("Missing header in region file!");
-        return null;
-      }
-      file.seek(4 * index);
-      int loc = file.readInt();
-      int numSectors = loc & 0xFF;
-      int sectorOffset = loc >> 8;
-      file.seek(SECTOR_SIZE + 4 * index);
-      int timestamp = file.readInt();
-      if (length < sectorOffset * SECTOR_SIZE + 4) {
-        Log.warnf("Chunk %s is outside of region file %s! Expected chunk data at offset %d but file length is %d.%n", chunkPos, regionFile.getName(), sectorOffset * SECTOR_SIZE, length);
-        return null;
-      }
-      file.seek(sectorOffset * SECTOR_SIZE);
+  private static ChunkDataSource readChunkData(RandomAccessFile file, ChunkPosition chunkPos) throws IOException {
+    long index = getMCAChunkIndex(chunkPos);
 
-      int chunkSize = file.readInt();
+    long length = file.length();
+    if (length < SECTOR_SIZE << 1) {
```

Comment:
I think it's clearer to use a `* 2` here, though it doesn't really matter. Adding a comment for a brief explanation

```suggestion
    // header is 2 sectors long, one for the offset+size, the other for timestamps
    if (length < SECTOR_SIZE * 2) {
```

Reply:
```suggestion
    // header is 2 sectors long: location table + timestamp table
    if (length < 2 * SECTOR_SIZE) {
```